### ISSUE TITLE
Plumbing Machines get stuck forever on small units of chemicals, requesting impossibly small amounts of reagents.

### ISSUE THREAD

**Issue by mGuv** [tgstation/tgstation#58993]
I found this playing offline after encountering it many times online.

## Reproduction:

- Place Three Synthesizers
- Set them to all output the same chemical (I use Bromine)
- Wrench all three Synthesizers
- Connect them all to the same Pipe
- Place a Reaction chamber that when wrenched would connect the Red Input pipe to the same pipes as the Synthesizers
- Set the Reaction chamber to ask for 1 unit of Bromine
- Wrench the Chamber in
- Observe that the Chamber never ends the "Filling" stage and gets stuck on 0.99 of Bromine.

## Deep Dive

So I checked out the code and ran this scenario locally and found the following observations.

- The chamber starts by asking for 1u of Bromine
- The transfer handling code looks for any machines that can provide Bromine and finds the three chambers
- It then tries to pull equally from these chambers, by asking each one for `0.32999998` of Bromine, which I imagine is rounded to 0.33 Bromine
- At the end of the update cycle, it's at 0.99 units of Brominne
- Next update cycle, it's short 0.01 units of Bromine and repeats the above... However, 0.01/3 then rounded becomes 0
- The chamber fails to pull the last 0.01 as a result and is now stuck forever "filling"

## Note

This can happen in MANY scenarios. From having multiple sources of the same, to machines pulling the same chems from an output that generates chems over time, to impurities in reactions causing fractionable outputs. Many of my chem set ups end up deadlocked due to requesting tiny fractions.

**Comment by mGuv** [comment]
I have since been able to make 100% pure Cyanide, and it indeed does come out at 1u, rather than `0.9999999994` and works around the iusse.

However, I have since had even perfect ratio plumbing machines get stuck due to the same problem. It seems anything that generates/pulls in odd ratios per tick (so synthesizers, reaction chambers, acclimators, etc.) can end up in this situation should the math align.

**Comment by mGuv** [comment]
I have updated the original post with a more easy to follow example. I am able to replicate this on demand very easily.

**Comment by mGuv** [comment]
@Time-Green, I have been told to get your attention to look at this?

**Comment by mGuv** [comment]
So as the issue is to do with rounding and float math. I wondered: why are we even rounding? There seems to be no reason to it other than to AVOID floating point issues. Rounding is hugely impacted by floating point issues.

Out of curiosity, I removed all the rounding in the chemistry code and tried all sorts of reactions to see if it caused any issues and... It behaves exactly how it did before and doesn't have the stuck scenarios.

Even my complicated setup of automating a ton of meds works as it did before.

![image](https://user-images.githubusercontent.com/7647687/118266481-93190180-b4b2-11eb-8ece-c3818eb382fc.png)

**Comment by Time-Green** [comment]
> So as the issue is to do with rounding and float math. I wondered: why are we even rounding? There seems to be no reason to it other than to AVOID floating point issues. Rounding is hugely impacted by floating point issues, the less precision you try to store, the more likely you are going to end up with a number that can't be accurately represented under the hood.
> 
> Out of curiosity, I removed all the rounding in the chemistry code to do with plumbing/handling and tried all sorts of reactions to see if it caused any issues and... It behaves exactly how it did before and doesn't have the stuck scenarios.
> 
> Even my complicated setup of automating a ton of meds works as it did before.
> 
> ![image](https://user-images.githubusercontent.com/7647687/118266481-93190180-b4b2-11eb-8ece-c3818eb382fc.png)
> 
> So again, why round? Float math works best when left alone.

Floats frequently ended up with amounts like 49.99999 because the components were floaterized to amounts like 16,6673829 and wouldn’t end up becoming 50, but dumb stuff like 50.0001. This would only escalate with factory size and ended up forcing players to frequently plunger all the machinery.

The underlying issue may have been fixed since you didn’t run into it. I’m fine with you PRing the change, you actually made a really great analyses and I don’t see the harm in tryinf. If plumbing blows up internally again it can always be reverted

**Comment by mGuv** [comment]
> Floats frequently ended up with amounts like 49.99999 because the components were floaterized to amounts like 16,6673829 and wouldn’t end up becoming 50, but dumb stuff like 50.0001. This would only escalate with factory size and ended up forcing players to frequently plunger all the machinery.
> 
> The underlying issue may have been fixed since you didn’t run into it. I’m fine with you PRing the change, you actually made a really great analyses and I don’t see the harm in tryinf. If plumbing blows up internally again it can always be reverted

Damn, @Time-Green, I see what you mean. As a test I made a chamber pull 99 units of Hydrogen and 33 units of Nitrogen. It resulted in 98.9998 Ammonia. Upon then pulling it in parts of 9 ammonia, the final 11th iteration only got 8.9998 and was stuck on "Filling".

I added more ammonia to the pipes and it was STILL stuck. Which I didn't expect. So I dug deeper and found this:
```
if(amount <= CHEMICAL_QUANTISATION_LEVEL)//To prevent small amount problems.
    return FALSE
```

Which ironically is causing the machine to get stuck. However, this is part of `/datum/reagents/proc/add_reagent` which I am nervous to change as it seems to be part of the backbone of all of chemistry.

I removed that statement and ran the same tests and everything seems fine. It never made `98.9998` and instead made `99`. And then split equally in to 11 equals parts of 9.

 Do you have any other examples of things that used to break?

**Comment by Time-Green** [comment]
> > Floats frequently ended up with amounts like 49.99999 because the components were floaterized to amounts like 16,6673829 and wouldn’t end up becoming 50, but dumb stuff like 50.0001. This would only escalate with factory size and ended up forcing players to frequently plunger all the machinery.
> > The underlying issue may have been fixed since you didn’t run into it. I’m fine with you PRing the change, you actually made a really great analyses and I don’t see the harm in tryinf. If plumbing blows up internally again it can always be reverted
> 
> Damn, @Time-Green, I see what you mean. As a test I made a chamber pull 99 units of Hydrogen and 33 units of Nitrogen. It resulted in `98.9998` Ammonia. Upon then pulling it in parts of 9 ammonia, the final 11th iteration only got `8.9998` and was stuck on "Filling".
> 
> I added more ammonia to the pipes and it was STILL stuck. Which I didn't expect. So I dug deeper and found this:
> 
> ```
> if(amount <= CHEMICAL_QUANTISATION_LEVEL)//To prevent small amount problems.
>     return FALSE
> ```
> 
> Which ironically is causing the machine to get stuck. As it's asking for a smaller unit than that - so it gets rejected and asks for it again, over and over. However, this is part of `/datum/reagents/proc/add_reagent` which I am nervous to change as it seems to be part of the backbone of all of chemistry.
> 
> I removed that statement and ran the same tests and everything seems fine. It never made `98.9998` and instead made `99`. And then split equally in to 11 equals parts of 9.
> 
> Do you have any other examples of things that used to break?

No, I think that was it for the floating point errots. Nice work! That’s been plagueing plumbing since the beginning of time

**Comment by mGuv** [comment]
Upon further testing and I'm unsure how I did it. I ended up with an actual floating point math issue =/

It's trying to do:

`98.99999 + 0.0000025431316` and the answer is `98.99999` due to the way floats work. It can't represent that tiny change. Which is why I assume there's a load of rounding/small unit handling code. Hmm. Need to rethink my approach

**Comment by mGuv** [comment]
But rounding doesn't work either, as in the current implementation.

Rounding to `0.01` makes the assumption that every number of two decimal places exist. That's not how float works.

`round(1 / 3, 0.01` does not equal `0.33`, as `0.33` does not exist as a float. The closest number (and what comes out of the round function) is `0.32999998`.

That tiny discrepancy between `0.33` and `0.32999998` is what causes issues down the line - as a target volume is never reached due due to that missing  `0.00000002`, which it will eventually try and ask for: but the rounding code kills it to 0.

**Comment by Time-Green** [comment]
Fixed in #59131
